appcelerator/titanium_mobile
# [TIMOB-13642] Android: Fixed bug where "relative" image paths won't load

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-13642

**Summary:**
- Relative image path was broken due to extra slash added to path, which Android asset manager could not handle.
  * Solved by trimming extra slashes.
  * Not an issue with external files since filesystem can handle extra slashes.
- Fixed resolution of paths containing `"../"` when loading an image from a JS file in a subdirectory.
- Added support for "content://" image URLs. (Used to only work via blobs.)
- Added support for "android.resource://" image URLs. (Used to only work via blobs.)
  * Allows `Ti.Filesystem.getResRawDirectory() + "MyImage"` to work.
  * Allows `"android.resource://" + Ti.App.id + "/drawable/MyImage"` to work.

**Note:**
Relative paths are currently handled differently between platforms.
- On Android, paths are relative to the JavaScript file.
- On iOS and Windows, paths are relative to the app's "Resources" directory.

I have not changed this behavior since that would be a breaking-change.

**Test:**
1. Create a "Classic App" project.
2. Replace the project's "app.js" with the code below. _(This code assumes the project has a "tab1.png".)_
3. Build and run the app on Android.
4. Scroll through the app, verifying that every black box has loaded an image.

```javascript
function createImageGroupFor(path) {
	var container = Ti.UI.createView({
		layout: "vertical",
		top: "5dp",
		left: "5dp",
		right: "5dp",
		backgroundColor: "#222222",
		width: Ti.UI.FILL,
		height: Ti.UI.SIZE,
	});
	container.add(Ti.UI.createLabel({
		text: path,
		color: "#CCCCCC",
		top: "5dp",
		left: "5dp",
		right: "5dp",
	}));
	container.add(Ti.UI.createImageView({
		image: path,
		preventDefaultImage: true,
		bottom: "5dp",
		width: "80dp",
		height: "80dp",
	}));
	return container;
}

var window = Ti.UI.createWindow({ fullscreen: true });
var scrollView = Ti.UI.createScrollView({
	layout: "vertical",
	scrollType: "vertical",
	showHorizontalScrollIndicator: false,
	shorVerticalScrollIndicator: true,
	backgroundColor: "gray",
});
scrollView.add(createImageGroupFor("/assets/images/tab1.png"));
scrollView.add(createImageGroupFor("assets/images/tab1.png"));
scrollView.add(createImageGroupFor("./assets/images/tab1.png"));
scrollView.add(createImageGroupFor("././assets/images/.././images/tab1.png"));
scrollView.add(createImageGroupFor(Ti.Filesystem.getFile("assets/images/tab1.png").nativePath));
if (Ti.Platform.name === "android") {
	var externalFilePath = Ti.Filesystem.applicationDataDirectory + "copy.png";
	Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, "assets/images/tab1.png").copy(externalFilePath);
	scrollView.add(createImageGroupFor(externalFilePath));
	externalFilePath = Ti.Filesystem.getFile(externalFilePath).nativePath;
	scrollView.add(createImageGroupFor(externalFilePath));
	externalFilePath = externalFilePath.substring("file:///".length);
	scrollView.add(createImageGroupFor("content://" + Ti.App.id + ".tifileprovider/filesystem/" + externalFilePath));
	scrollView.add(createImageGroupFor(Ti.Filesystem.resourcesDirectory + "assets/images/tab1.png"));
	scrollView.add(createImageGroupFor("file:///android_asset/Resources/assets/images/tab1.png"));
	scrollView.add(createImageGroupFor("android.resource://" + Ti.App.id + "/drawable/appicon"));
}
scrollView.add(Ti.UI.createView({ width: Ti.UI.FILL, height: "5dp" }));
window.add(scrollView);
window.open();
```
